src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php

```diff
@@ -66,4 +71,86 @@ public function log($level, $message, array $context = []): void
+     */
+    public function testHttp2PushVulcain()
+    {
+        if (\PHP_VERSION_ID >= 70300 && \PHP_VERSION_ID < 70304) {
+            $this->markTestSkipped('PHP 7.3.0 to 7.3.3 don\'t support HTTP/2 PUSH');
+        }
+
+        if (!\defined('CURLMOPT_PUSHFUNCTION') || 0x073d00 > ($v = curl_version())['version_number'] || !(CURL_VERSION_HTTP2 & $v['features'])) {
+            $this->markTestSkipped('curl <7.61 is used or it is not compiled with support for HTTP/2 PUSH');
+        }
+
+        $this->startVulcain();
+
+        $logger = new class() extends AbstractLogger {
+            public $logs = [];
+
+            public function log($level, $message, array $context = []): void
+            {
+                $this->logs[] = $message;
+            }
+        };
+
+        $client = new CurlHttpClient(['verify_peer' => false, 'verify_host' => false]);
+        $client->setLogger($logger);
+
+        $responseAsArray = $client->request('GET', 'https://127.0.0.1:3000/json', [
+            'headers' => [
+                'Preload' => '/documents/*/id',
+            ]
+        ])->toArray();
+        foreach ($responseAsArray['documents'] as $document) {
+            $docAsArray = $client->request('GET', 'https://127.0.0.1:3000' . $document['id'])->toArray();
+        }
+
+        $expected = [
+            'Request: "GET https://127.0.0.1:3000/json"',
+            'Queueing pushed response: "https://127.0.0.1:3000/json/1"',
+            'Queueing pushed response: "https://127.0.0.1:3000/json/2"',
+            'Queueing pushed response: "https://127.0.0.1:3000/json/3"',
+            'Response: "200 https://127.0.0.1:3000/json"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/1"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/2"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/3"',
+            'Accepting pushed response: "GET https://127.0.0.1:3000/json/1"',
+            'Response: "200 https://127.0.0.1:3000/json/1"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/2"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/3"',
+            'Accepting pushed response: "GET https://127.0.0.1:3000/json/2"',
+            'Response: "200 https://127.0.0.1:3000/json/2"',
+            'Unused pushed response: "https://127.0.0.1:3000/json/3"',
+            'Accepting pushed response: "GET https://127.0.0.1:3000/json/3"',
+            'Response: "200 https://127.0.0.1:3000/json/3"',
+        ];
+        $this->assertSame($expected, $logger->logs);
+    }
+
+    private function startVulcain()
+    {
+        $process = new Process(['vulcain'], null, [
+            'DEBUG' => 1,
+            'UPSTREAM' => 'http://127.0.0.1:8057',
+            'ADDR' => ':3000',
+            'KEY_FILE' => 'src/Symfony/Contracts/HttpClient/Test/Fixtures/tls/server.key',
+            'CERT_FILE'=> 'src/Symfony/Contracts/HttpClient/Test/Fixtures/tls/server.crt',
```

Comment:
these 2 files could (should) be in the component isn't it?

Reply:
Wep :+1: I moved them